spec/tree-sitter-language-mode-spec.js

```diff
@@ -766,6 +766,61 @@ describe('TreeSitterLanguageMode', () => {
         ])
       })
 
+      it('terminates comment token at the end of an injection, so that the next injection is NOT a continuation of the comment', async () => {
+        const ejsGrammar = new TreeSitterGrammar(
+          atom.grammars,
+          ejsGrammarPath,
+          {
+            id: 'ejs',
+            parser: 'tree-sitter-embedded-template',
+            scopes: {
+              '"<%"': 'directive',
+              '"%>"': 'directive'
+            },
+            injectionPoints: [
+              {
+                type: 'template',
+                language (node) { return 'javascript' },
+                content (node) { return node.descendantsOfType('code') },
+                newlinesBetween: true
+              },
+              {
+                type: 'template',
+                language (node) { return 'html' },
+                content (node) { return node.descendantsOfType('content') }
+              }
+            ]
+          }
+        )
+
+        atom.grammars.addGrammar(jsGrammar)
+        atom.grammars.addGrammar(htmlGrammar)
+
+        buffer.setText('<% // js comment %>\n<% b() %>')
+        const languageMode = new TreeSitterLanguageMode({
+          buffer,
+          grammar: ejsGrammar,
+          grammars: atom.grammars
+        })
+        buffer.setLanguageMode(languageMode)
+
+        expectTokensToEqual(editor, [
+          [
+            { text: '<%', scopes: ['directive'] },
+            { text: ' // js comment ', scopes: [] },
+            { text: '%>', scopes: ['directive'] },
+            { text: '', scopes: ['html'] }
+          ],
+          [
+            { text: '<%', scopes: ['directive'] },
+            { text: ' ', scopes: [] },
+            { text: 'b', scopes: ['function'] },
+            { text: '() ', scopes: [] },
+            { text: '%>', scopes: ['directive'] }
+          ]
+        ])
```

Comment:
Prior to the changes to `src/tree-sitter-language-mode.js` introduced in this pull request, Atom would produce the following tokens when executing this test:

```javascript
[
  [
    { text: '<%', scopes: ['directive'] },
    { text: ' // js comment ', scopes: [] },
    { text: '%>', scopes: ['directive'] },
    { text: '', scopes: ['html'] }
  ],
  [
    { text: '<%', scopes: ['directive'] },
    { text: ' b() ', scopes: [] },
    { text: '%>', scopes: ['directive'] }
  ]
]
```

Note that in this tokenization ☝️, `b()` is not tokenized as a function. Instead, `b()` is treated as a continuation of the comment on the preceding line:

```
<% // js comment %>
<% b() %>
```

The enhancements in this pull request (specifically ef94657438f4b3a70f0e0a7cc7895939626f1c72) allow Atom to correctly identify that the JavaScript comment is implicitly terminated at the end of the first line, and the second line begins a new JavaScript statement.

Reply:
I think it might help tell the story if you added a scope mapping for comment tokens in the `jsGrammar` defined above in the `beforeEach`:

```js
jsGrammar = new TreeSitterGrammar(atom.grammars, jsGrammarPath, {

  // ...

  scopes: {

    // ...

    'comment': 'comment'
  }
})
```

That way, you'd see a `comment` token in the expected output, something like:

```js
{ text: '<%', scopes: ['directive'] },
{ text: ' ', scopes: [] },
{ text: '// js comment ', scopes: ['comment'] },
{ text: '%>', scopes: ['directive'] },

```

Reply:
Sweet! Thanks, @maxbrunsfeld. I explored something similar this morning, but couldn't get it to cooperate for some reason. Thanks for pointing me in the right direction.

f18ad1eea191821bc26ee4ba24cb5e7eb5f4b92d implements this suggestion.